cpp/src/lists/contains.cu

```diff
@@ -105,48 +140,168 @@ struct search_list_fn {
     // If the key is found, return its found position in the list from `found_iter`.
     return found_iter == end ? NOT_FOUND_SENTINEL : *found_iter;
   }
+};
+
+/**
+ * @brief Functor to perform searching for index of a key element in a given list, specialized
+ * for nested types.
+ */
+template <typename KeyValidityIter, typename EqComparator>
+struct search_list_nested_types_fn {
+  duplicate_find_option const find_option;
+  KeyValidityIter const key_validity_iter;
+  EqComparator const d_comp;
+  bool const search_key_is_scalar;
+
+  search_list_nested_types_fn(duplicate_find_option const find_option,
+                              KeyValidityIter const& key_validity_iter,
+                              EqComparator const& d_comp,
+                              bool search_key_is_scalar)
+    : find_option(find_option),
+      key_validity_iter(key_validity_iter),
+      d_comp(d_comp),
+      search_key_is_scalar(search_key_is_scalar)
+  {
+  }
+
+  __device__ size_type operator()(list_device_view list) const
+  {
+    // A null list or null key will result in a null output row.
+    if (list.is_null() || !key_validity_iter[list.row_index()]) { return NULL_SENTINEL; }
+
+    return find_option == duplicate_find_option::FIND_FIRST ? search_list<true>(list)
+                                                            : search_list<false>(list);
+  }
 
-  /**
-   * @brief Return a pair of index iterators {begin, end} to loop through elements within a list.
-   *
-   * Depending on the value of `forward`, a pair of forward or reverse iterators will be
-   * returned, allowing to loop through elements in the list in first-to-last or last-to-first
-   * order.
-   *
-   * Note that the element indices always restart to `0` at the first position in each list.
-   *
-   * @tparam forward A boolean value indicating whether we want to iterate elements in the list by
-   *         forward or reverse order.
-   * @param size The number of elements in the list.
-   * @return A pair of {begin, end} iterators to iterate through the range `[0, size)`.
-   */
+ private:
   template <bool forward>
-  static __device__ auto element_index_pair_iter(size_type const size)
+  __device__ inline size_type search_list(list_device_view list) const
   {
-    if constexpr (forward) {
-      return thrust::pair(thrust::make_counting_iterator(0), thrust::make_counting_iterator(size));
+    using cudf::experimental::row::lhs_index_type;
+    using cudf::experimental::row::rhs_index_type;
+
+    auto const [begin, end] = element_index_pair_iter<forward>(list.size());
+    auto const found_iter =
+      thrust::find_if(thrust::seq, begin, end, [&] __device__(auto const idx) {
```

Comment:
I'm not sure if we want to capture everything by reference here. The `list_device_view` should probably be passed by value unless you know of a reason not to do so.
```suggestion
      thrust::find_if(thrust::seq, begin, end, [list, &d_comp, &search_key_is_scalar] __device__(auto const idx) {
```